posts/2023-06-07-kotlin-ktor-keycloak.adoc

```diff
@@ -0,0 +1,357 @@
+
+.Configure a Slack App worskpace
+image::slack_create_app_2.png[Diagram, 450]
+
+To finalize the App creation you need to set some permissions:
+
+.Setup Slack App permissions
+image::slack_app_permissions.png[Diagram, 400]
+
+In `Basic Information` you can personalize your app appearance, but for now let's jump to `OAuth & Permission` and, from there, scroll down to the `Scopes` section. In `User Token Scopes` add the scopes `chat:write`:
+
+.Setup User Token Scopes
+image::slack_app_scopes.png[Diagram, 500]
+
+Then scroll up and click `Install to Workspace`. That will take you to a permissions screen: click `Allow`. You can see that a `User OAuth Token` was created as well.
+
+.Slack App User OAuth token
+image::slack_user_oauth_token.png[Diagram, 500]
+
+We will need this `User OAuth Token` to be able to send messages to Slack via this App.
+
+=== Put it all together
+
+We can now send that `luna-vacation-api` response to a Slack channel. The channel I'll use is the chat with myself.
+In Slack, channel ids can be seen when you select a channel or a direct message chat
+and then click on the top details of the channel/person. At the very bottom you can see "Channel ID".
+
+For the communication with Slack we will use Slack's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[Bolt SDK],
+starting by adding the necessary dependencies:
+
+[source,kotlin]
+----
+implementation("com.slack.api:bolt:1.29.2")
+implementation("com.slack.api:bolt-servlet:1.29.2")
+implementation("com.slack.api:bolt-jetty:1.29.2")
+implementation("org.slf4j:slf4j-simple:1.7.36")
+----
+
+Now let's https://slack.dev/java-slack-sdk/guides/getting-started-with-bolt[compose our message].
+The function `getIfImOnVacation` is wrapping the `luna-vacation-api` request.
+
+[source,kotlin]
+----
+val text = getIfImOnVacation(client)
+val response = slack.methods("xoxp-...").chatPostMessage { req: ChatPostMessageRequestBuilder ->
+    req
+        .channel("<replace by my channel>")
+        .text(text)
+}
+----
+
+Make sure `luna-vacation-api` is still running and now let's run `luna-vacation-bot`. If you check Slack you should see a message:
+
+.Slack message
+image::slack_message.png[Diagram, 300]
+
+The complete source code for the https://github.com/LeonorLunatech/luna-vacation-api[luna-vacation-api] and https://github.com/LeonorLunatech/luna-vacation-bot[luna-vacation-bot] are available in GitHub.
+
+=== Next steps
+
+We can transform this kind of application integration in a real automation by running it periodically.
+A cron job allows to achieve this goal and the http://www.quartz-scheduler.org/documentation/quartz-2.3.0/quick-start.html[quartz-scheduler] library integrates well with Kotlin.
+The cron job can be configured as often was one wishes like several time a day or once a month, for example.
+
+In this post we used a completely dummy API. For a real use case we would replace it by an API that provides data reports, for example.
```

Comment:
```suggestion
In this blog post we used a dummy API. For a real use case we could replace it with an API that provides data reports, for example.
```
Not sure the word `completely` is needed. `Dummy` is good enough I think.

Reply:
My mediterranean blood.. exaggerating things a bit 😄